Updated preview API to add Kernel Config

srivatsasinha opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/b321e1dc-5e07-4940-9cd6-a65ac6729a8b)

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [ ] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [x] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

Commit:
Updated preview API to add Kernel Config and LRO Headers

Commit:
Fixing formatting issue

MaggieHust commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/examples/Compute/createOrUpdate/ComputeInstance.json:
"displayName": "test"

could this part be rename to something that is more meaningful?

srivatsasinha commented:
"Swagger LintDiff" is failing. Its asking to put Location header in the LRO API. But we are already returning Azure-AsyncOperation which serves the same purpose. Please look into this.

mentat9 commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:
```diff
@@ -3174,6 +3196,29 @@
+          "description": "Language of the kernel"
+        }
+      },
+      "additionalProperties": true
```
>additionalProperties [](http://example.com/codeflow?start=7&length=20)

Try to avoid this pattern wherever possible. Untyped/unschematized properties are difficult and dangerous for clients to consume. Generally, this pattern should only be used when needed and the object properties are dynamic, unknown, or user defined. Ref: [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined).

srivatsasinha replied:
Currently we are using this properties bag to extend new features in private preview. We will consider removing it during GA

visingla-ms replied:
I would suggest removing it now and monitor your preview requirements to release newer preview version with more specific properties. Using additonalProperties is not recommended for this purpose.

mentat9 commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:
```diff
@@ -2978,6 +2996,10 @@
             "target"
           ]
+        },
+        "kernel": {
```
>kernel [](http://example.com/codeflow?start=9&length=6)

Consider naming this something more descriptive/specific (jupyterConfig, jupyterKernelConfig,or similar/better).

srivatsasinha replied:
We will consider renaming this to jupyterKernelConfig. 

@Ada, Please consider making changes to the C3 accordingly.

adrosa replied:
we are leaving as kernel. this value has been in production for quite a while, we are just adding documentation. due to the design changing the name of values like this causes issues because it is stateful on customer controlled compute.

srivatsasinha replied:
Had a discussion on this, currently only jupyter is supported but we will plan for other kernels too so in line we wish to keep the name same.

Commit:
Updated as per comments

Commit:
Updated as per comments

Commit:
Reverting changes

Commit:
Reverted header changes

srivatsasinha commented:
@mentat9 , I have reverted some header changes as it came to my attention that someone has already worked on same, so i am avoiding duplication. I have addressed rest of the comments and all required check are passing. Please review

TimLovellSmith commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:
```diff
@@ -3174,6 +3178,29 @@
+        },
+        "language": {
+          "type": "string",
+          "description": "Language of the kernel"
```
>Language of the kernel" [](http://example.com/codeflow?start=26&length=23)

better docs - speicified how? what are valid values?

adrosa replied:
any value is valid, specifying as Enum is confusing.

anisoptera replied:
Yes, this is going to be a string in applications that want to display the kernel. It could be anything, including things we've never heard of, or even things that don't exist if the user wants to have an arbitrary division in the list or something.

TimLovellSmith commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:
```diff
@@ -3174,6 +3178,29 @@
+      "properties": {
+        "argv": {
+          "type": "array",
+          "description": "Argument to the the runtime",
```
>description": "Argument to the the runtime", [](http://example.com/codeflow?start=11&length=44)

warning: passing commandline arguments through REST APIs could cause security issues - are you sanitizing these?

adrosa replied:
not a concern given nature of service..

srivatsasinha replied:
This is as per kernel spec of jupyter. Hence these commandline options shouldn't be of concern https://jupyter-client.readthedocs.io/en/stable/kernels.html

TimLovellSmith commented:
> "Swagger LintDiff" is failing. Its asking to put Location header in the LRO API. But we are already returning Azure-AsyncOperation which serves the same purpose. Please look into this.

HTTP requires a Location header with 202 Accepted

TimLovellSmith commented:
>           },

202 shouldn't indicate complete success - this looks incomplete, since you've declared it as 202 x-ms-long-running-operation the intention looks to be to provide a Location header to track the async operation

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:715 in 87997da. [](commit_id = 87997dad1bf6e0c5be1f870e71cb878b786d531d, deletion_comment = False)

TimLovellSmith commented:
>           },

202 shouldn't indicate complete success - this looks incomplete, since you've declared it as 202 x-ms-long-running-operation the intention looks to be to provide a Location header to track the async operation - along with a final status code when the operation completes

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:759 in 87997da. [](commit_id = 87997dad1bf6e0c5be1f870e71cb878b786d531d, deletion_comment = False)

TimLovellSmith commented:
>           },

same - 202 shouldn't indicate complete success - this looks incomplete, since you've declared it as 202 x-ms-long-running-operation the intention looks to be to provide a Location header to track the async operation

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:803 in 87997da. [](commit_id = 87997dad1bf6e0c5be1f870e71cb878b786d531d, deletion_comment = False)

TimLovellSmith commented:
>           },

Since these aren't part of your main changes, maybe you'd like to address it in a follow-up PR

---
In reply to: [1718344872](https://github.com/Azure/azure-rest-api-specs/pull/25361#issuecomment-1718344872) [](http://example.com/codeflow?ancestors=1718344872)

---
Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:715 in 87997da. [](commit_id = 87997dad1bf6e0c5be1f870e71cb878b786d531d, deletion_comment = False)

TimLovellSmith commented:
You're getting many breaking change warnings because this preview version doesn't looks like it has feature parity with 2023-04-01, but I guess most of those changes date back to 2023-06-01-preview([main](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-06-01-preview/machineLearningServices.json)

Commit:
Reverting change

Commit:
Reverting example

srivatsasinha commented:
> You're getting many breaking change warnings because this preview version doesn't looks like it has feature parity with 2023-04-01, but I guess most of those changes date back to 2023-06-01-preview([main](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-06-01-preview/machineLearningServices.json)

I dont think we can fix it here. We will inform our team swagger coordinators.

srivatsasinha commented:
> > ```
> >       },
> > ```
> 
> Since these aren't part of your main changes, maybe you'd like to address it in a follow-up PR
> 
> In reply to: [1718344872](https://github.com/Azure/azure-rest-api-specs/pull/25361#issuecomment-1718344872)
> 
> Refers to: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-08-01-preview/machineLearningServices.json:715 in [87997da](https://github.com/Azure/azure-rest-api-specs/commit/87997dad1bf6e0c5be1f870e71cb878b786d531d). [](commit_id = [87997da](https://github.com/Azure/azure-rest-api-specs/commit/87997dad1bf6e0c5be1f870e71cb878b786d531d), deletion_comment = False)

These are currently being handled by a separate team, that's why i reverted my changes.

rkmanda commented:
> > You're getting many breaking change warnings because this preview version doesn't looks like it has feature parity with 2023-04-01, but I guess most of those changes date back to 2023-06-01-preview([main](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2023-06-01-preview/machineLearningServices.json)
> 
> I dont think we can fix it here. We will inform our team swagger coordinators.

Its not clear why these cant be fixed here. Can you elaborate? Getting breaking change approval is not optional.

rkmanda commented:
Please do the needful for step 1 to continue the review

Commit:
Added version parameter and removed addition properties as per guidelines

IntelUnnati commented:
/pr RequestMerge
